From linux-kernel@vger.kernel.org Fri Jan  9 21:59:44 2009
X-Envelope-From: <git-commits-head-owner@vger.kernel.org>
X-Envelope-To: <vger@sl-h.de>
X-Delivery-Time: 1231534791
X-UID: 400146
Return-Path: <git-commits-head-owner@vger.kernel.org>
X-RZG-CLASS-ID: mi
Received: from vger.kernel.org ([209.132.176.167])
	by mailin.webmailer.de (zeb mi19) (RZmta 18.7)
	with ESMTP id R013bdl09Kucei for <vger@sl-h.de>;
	Fri, 9 Jan 2009 21:59:50 +0100 (MET)
Received: (majordomo@vger.kernel.org) by vger.kernel.org via listexpand
	id S1756210AbZAIU7s (ORCPT <rfc822;vger@sl-h.de>);
	Fri, 9 Jan 2009 15:59:48 -0500
Received: (majordomo@vger.kernel.org) by vger.kernel.org id S1754402AbZAIU7s
	(ORCPT <rfc822;git-commits-head-outgoing>);
	Fri, 9 Jan 2009 15:59:48 -0500
Received: from hera.kernel.org ([140.211.167.34]:38493 "EHLO hera.kernel.org"
	rhost-flags-OK-OK-OK-OK) by vger.kernel.org with ESMTP
	id S1756740AbZAIU7r (ORCPT
	<rfc822;git-commits-head@vger.kernel.org>);
	Fri, 9 Jan 2009 15:59:47 -0500
Received: from hera.kernel.org (IDENT:U2FsdGVkX1+i0E9z1rFfPU+gXLQOjhtmns1Dtpr6fxs@localhost [127.0.0.1])
	by hera.kernel.org (8.14.2/8.14.2) with ESMTP id n09KxiS4014624
	(version=TLSv1/SSLv3 cipher=DHE-RSA-AES256-SHA bits=256 verify=NO)
	for <git-commits-head@vger.kernel.org>; Fri, 9 Jan 2009 20:59:44 GMT
Received: (from dwmw2@localhost)
	by hera.kernel.org (8.14.2/8.13.1/Submit) id n09Kxi3F014606
	for git-commits-head@vger.kernel.org; Fri, 9 Jan 2009 20:59:44 GMT
Date:	Fri, 9 Jan 2009 20:59:44 GMT
Message-Id: <200901092059.n09Kxi3F014606@hera.kernel.org>
From:	Linux Kernel Mailing List <linux-kernel@vger.kernel.org>
To: git-commits-head@vger.kernel.org
Subject: ioat: fix self test for multi-channel case
MIME-Version: 1.0
Content-Type: text/plain;
  charset=UTF-8
Content-Transfer-Encoding: 8bit
X-Git-Commit: b9bdcbba010c2e49c8f837ea7a49fe006b636f41
X-Git-Parent: 652afc27b26859a0ea5f6db681d80b83d2c43cf8
X-Virus-Scanned: ClamAV 0.93.3/8848/Fri Jan  9 13:16:38 2009 on hera.kernel.org
X-Virus-Status:	Clean
X-Spam-Status: No, score=-2.6 required=5.0 tests=AWL,BAYES_00,
	UNPARSEABLE_RELAY autolearn=ham version=3.2.5
X-Spam-Checker-Version:	SpamAssassin 3.2.5 (2008-06-10) on hera.kernel.org
X-Greylist: Sender IP whitelisted, not delayed by milter-greylist-4.0 (hera.kernel.org [127.0.0.1]); Fri, 09 Jan 2009 20:59:46 +0000 (UTC)
Sender:	git-commits-head-owner@vger.kernel.org
Precedence: bulk
List-ID: <git-commits-head.vger.kernel.org>
X-Mailing-List:	git-commits-head@vger.kernel.org
Status: R
X-Status: NC
X-KMail-EncryptionState:  
X-KMail-SignatureState:  
X-KMail-MDN-Sent:  

Gitweb:     http://git.kernel.org/git/?p=linux/kernel/git/torvalds/linux-2.6.git;a=commit;h=b9bdcbba010c2e49c8f837ea7a49fe006b636f41
Commit:     b9bdcbba010c2e49c8f837ea7a49fe006b636f41
Parent:     652afc27b26859a0ea5f6db681d80b83d2c43cf8
Author:     Dan Williams <dan.j.williams@intel.com>
AuthorDate: Tue Jan 6 11:38:22 2009 -0700
Committer:  Dan Williams <dan.j.williams@intel.com>
CommitDate: Tue Jan 6 11:38:22 2009 -0700

    ioat: fix self test for multi-channel case
    
    In the multiple device case we need to re-arm the completion and protect
    against concurrent self-tests.  The printk from the test callback is
    removed as it can arbitrarily delay completion of the test.
    
    Cc: <stable@kernel.org>
    Cc: Maciej Sosnowski <maciej.sosnowski@intel.com>
    Signed-off-by: Dan Williams <dan.j.williams@intel.com>
---
 drivers/dma/ioat_dma.c |   13 +++++++------
 1 files changed, 7 insertions(+), 6 deletions(-)

diff --git a/drivers/dma/ioat_dma.c b/drivers/dma/ioat_dma.c
index e42e1ae..b3759c4 100644
--- a/drivers/dma/ioat_dma.c
+++ b/drivers/dma/ioat_dma.c
@@ -1340,12 +1340,11 @@ static void ioat_dma_start_null_desc(struct ioat_dma_chan *ioat_chan)
  */
 #define IOAT_TEST_SIZE 2000
 
-DECLARE_COMPLETION(test_completion);
 static void ioat_dma_test_callback(void *dma_async_param)
 {
-	printk(KERN_ERR "ioatdma: ioat_dma_test_callback(%p)\n",
-		dma_async_param);
-	complete(&test_completion);
+	struct completion *cmp = dma_async_param;
+
+	complete(cmp);
 }
 
 /**
@@ -1362,6 +1361,7 @@ static int ioat_dma_self_test(struct ioatdma_device *device)
 	dma_addr_t dma_dest, dma_src;
 	dma_cookie_t cookie;
 	int err = 0;
+	struct completion cmp;
 
 	src = kzalloc(sizeof(u8) * IOAT_TEST_SIZE, GFP_KERNEL);
 	if (!src)
@@ -1401,8 +1401,9 @@ static int ioat_dma_self_test(struct ioatdma_device *device)
 	}
 
 	async_tx_ack(tx);
+	init_completion(&cmp);
 	tx->callback = ioat_dma_test_callback;
-	tx->callback_param = (void *)0x8086;
+	tx->callback_param = &cmp;
 	cookie = tx->tx_submit(tx);
 	if (cookie < 0) {
 		dev_err(&device->pdev->dev,
@@ -1412,7 +1413,7 @@ static int ioat_dma_self_test(struct ioatdma_device *device)
 	}
 	device->common.device_issue_pending(dma_chan);
 
-	wait_for_completion_timeout(&test_completion, msecs_to_jiffies(3000));
+	wait_for_completion_timeout(&cmp, msecs_to_jiffies(3000));
 
 	if (device->common.device_is_tx_complete(dma_chan, cookie, NULL, NULL)
 					!= DMA_SUCCESS) {
--
To unsubscribe from this list: send the line "unsubscribe git-commits-head" in
the body of a message to majordomo@vger.kernel.org
More majordomo info at  http://vger.kernel.org/majordomo-info.html

